src/main/java/com/cloudbees/jenkins/support/impl/SlaveCommandStatistics.java

```diff
@@ -0,0 +1,173 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.cloudbees.jenkins.support.impl;
+
+import com.cloudbees.jenkins.support.api.Component;
+import com.cloudbees.jenkins.support.api.Container;
+import com.cloudbees.jenkins.support.api.PrintedContent;
+import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import hudson.Extension;
+import hudson.ExtensionList;
+import hudson.FilePath;
+import hudson.Util;
+import hudson.model.Computer;
+import hudson.model.TaskListener;
+import hudson.remoting.Channel;
+import hudson.remoting.Command;
+import hudson.remoting.Request;
+import hudson.remoting.Response;
+import hudson.security.Permission;
+import hudson.slaves.ComputerListener;
+import java.io.File;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.regex.Pattern;
+import jenkins.model.Jenkins;
+
+@Extension
+public final class SlaveCommandStatistics extends Component {
+
+    private final Map<String, Statistics> statistics = Collections.synchronizedSortedMap(new TreeMap<>());
+
+    @Override
+    public String getDisplayName() {
+        return "Agent Command Statistics";
+    }
+
+    @Override
+    public Set<Permission> getRequiredPermissions() {
+        return Collections.singleton(Jenkins.ADMINISTER);
+    }
+
+    @Override
+    public void addContents(Container container) {
+        statistics.forEach((name, stats) -> container.add(new PrintedContent("nodes/slave/" + name + "/command-stats.md") {
+            @Override
+            protected void printTo(PrintWriter out) throws IOException {
+                stats.print(out);
+            }
+        }));
+    }
+
+    private static final class Statistics extends Channel.Listener {
+
+        /** Represents a tally of both the number of times some event occurred, and some integral metric associated with each event which should be summed. */
+        private static final class CountSum {
+            long count;
+            long sum;
+            void tally(long value) {
+                count++;
+                sum += value;
+            }
+            long count() {
+                return count;
+            }
+            long sum() {
+                return sum;
+            }
+        }
+        private final Map<String, CountSum> writes = new HashMap<>();
```

Comment:
It should be thread-safe, should not it? IIRC `Channel.Listener` may be invoked from different threads thanks from NIO

Reply:
Accesses are synchronized. See below.

Reply:
somehow missed that, sorry